content/blog/2019-11-07-immix/index.md

```diff
@@ -0,0 +1,210 @@
+this paper shows that immix can match state-of-the-art on three critical aspects: 
+*space efficiency*, *fast collection*, and *mutator performance*.
+
+# The Imperfection of Existing Tracing Collectors
+
+This paper focus on three important demands on garbage collectors:
+
+- *Space efficiency*: the less space overhead is, the better.
+- *Fast collection*: the faster the collector runs, the better.
+- *Mutator performance*: the faster the mutator runs, the better.
+
+This paper first investigates three existing canonical tracing collectors: 
+*semi-space* (SS), *mark-sweep* (MS), and *mark-compact* (MC). 
+
+![001](./001.png)
+
+The above figure shows the performance tradeoffs for canonical collectors. 
+It plots the geometric mean of total time, collection time, mutator time, 
+and mutator cache misses as a function of the heap size, 
+normalized to the best, for 20 DaCapo, SPECjvm98 and SPECjbb2000 benchmarks, 
+and shows 99% confidence intervals. 
+Figure (a) demonstrates that semi-space performs best when the heap size is large enough, 
+but sacrifices space efficiency due to the 2x space overhead; 
+Figure (b) shows that mark-compact does a bad job on achieving fast collection; 
+Figure (c) and (d) indicate that mark-sweep cannot compete with the other two in terms of mutator performance.
+
+Therefore, the primary goal of the new collector proposed by this paper is to guarantee all three demands.
+
+# Immix
+
+Immix contains two components: 
+a *mark-region* collector and an *opportunistic* defragmentation mechanism.
+
+### Mark-region
+
+A mark-region collector is similar to a mark-sweep collector, 
+with 2 differences: 
+(1) In mark-region, memory is divided into fixed-sized regions, 
+and objects may not span regions;
+(2) Instead of maintaining a free list, 
+which a mark-sweep collector allocates from and sweeps to, 
+a mark-region collector bumps allocates into free regions, 
+which are regions with no living objects, 
+and reclaims free regions during collection.
+
+A mark-region collector performs better on mutator performance than a mark-sweep collector because of its contiguous allocation nature. 
+However, it suffers from space-inefficiency as one single living object can hold a region unavailable. 
+Immix addresses this problem by 
+(1) refining this introduced mark-region collector to improve utilization of a region;
+(2) introducing a defragmentation mechanism, 
+which is presented in the next section.
+
+Immix refines mark-region by operating at two levels, 
+coarse-grained *blocks*, and fine-grained *lines*. 
+Comparing to the original mark-region collector, 
+the differences are: 
+(1) Objects may span lines; 
+(2) during reclamation,
+immix also identifies free lines in partially free blocks (formerly unavailable blocks); 
+(3) during allocation, 
+immix first tries to find contiguous free lines that are large enough for the allocates. 
+If it cannot find suitable lines, 
+then it allocates into free blocks.
```

Comment:
I considered this point, but I think introducing lines is mostly for lower fragmentation. In Table 4 (The last table I showed), the performance is not sensitive to line size. And they also say:

> Tuning line size seems most closely tied to the programming language’s influence on object size demographics, rather than the architecture.

You are definitely correct that the selection of block and line size is interesting and important. I'll add one bullet point in section "Implementation Details" to address it. Sounds good?

Reply:
OK, good point! Sounds good.